ovn-org/ovn-kubernetes
# [poc]  kubevirt: Add hypershift requirements to primary interface

**- What this PR does and why is it needed**
This is a draft PR to implement features needed for hypershift hosted workers:
- seamless kubevirt live migration
  - stable IPs
  - stable connections
- east/west and north/south communication
- access to services

# TODO
- [x] Don't set ip config at veth if kubevirt
- [x] Configure DHCP options at LSP
- [x] Use proxy_arp option from router LSP https://github.com/ovn-org/ovn/commit/8087cbc74628e785250f8c09e43f5e0e361e453e
  - https://github.com/ovn-org/ovn/pull/176
- [x] Add /32 routes ovn_cluster_router
- [x] Use route instead of policy for inbound traffic to specify the port.
- [ ] Sometimes new connections are not possible after live migration looks like endpoint slices are not correctly updated
  - https://github.com/kubernetes/kubernetes/pull/115907
  - backport: https://github.com/kubernetes/kubernetes/pull/116084
- [ ] Investigate `net.netfilter.nf_conntrack_tcp_loose=1` for snat + live migration
- [ ] Check that forcing the mac to the VM ends at ovn-k
  - [x] policy
  - [x] routes
  - [x] dhcp-options
- [x] Test services
- [x] Test connectivity + live migration
  - with proxy arp + LRP shared mac address we have < 1 second disruption
- [ ] Test postcopy migraiton
  - Ther is a 10s downtime, point to point route is correctly updated when annotation appears but traffic is not working.
  - Ednpoint slice for the new virt-launcher are not ready and not serving until src virt-launcher is not Completed
  - Traffic goes back between route change and virt-launcher src completed, but not at route change.
  - When annotation appears the src virt-launcher is not able to manage traffic
  - Maybe traffic is still going to the src virt-launcher when it already does not accept it.
  - Adding unpaused condition to target virt-launcher decreased 3-4 seconds the downtime
- [ ] Test network policies 
- [ ] Test hypershift
- [ ] Use kubevirt PRs:
  - https://github.com/kubevirt/kubevirt/pull/9144
  - https://github.com/kubevirt/kubevirt/pull/9193
  - https://github.com/kubevirt/kubevirt/pull/9290
  - https://github.com/kubevirt/kubevirt/pull/9330
- [ ] Use hypershift PRs:
  - https://github.com/openshift/hypershift/pull/2244
- [x] Investigate multi requested-chassis feature for live migration
  - It only works when LSP is not moved, meaning it only work with distributed switch.
- [ ] Reverse on to new ovnkube-master controllers, cluster-manager and network-controller-manager

# openshift issues
## "kube-dns" not found`
problem getting dns server `addLogicalPort failed for cluster1/virt-launcher-worker1-chdqx: failed composing DHCP options: services "kube-dns" not found`

For the poc we just fallback to openshift one, but we shoiuld "mimic" what kubevirt dhcp server is doing and read for virt-launcher pod resolv.conf, this can be done by the CNI annotating the pod then ovnkube-master will use those to generate DHCP options.

## node-valid-hostname.service
```
Failed Units: 1
  node-valid-hostname.service
```

## cannot migrate VMI: PVC test1-k7m4b-rhcos is not shared
```
You are using a client virtctl version that is different from the KubeVirt version running in the cluster
Client Version: v0.58.0
Server Version: v0.59.0-rc.1-19-ga5e3a1b9b
Error migrating VirtualMachine Internal error occurred: admission webhook "migration-create-validator.kubevirt.io" denied the request: Cannot migrate VMI, Reason: DisksNotLiveMigratable, Message: cannot migrate VMI: PVC test1-k7m4b-rhcos is not shared, live migration requires that all PVCs must be shared (using ReadWriteMany access mode)
```

We have to use the patch from other hypershift pocs to change the mode


We have to pass "hostname" same as kubevirt DHCP server and set it with the vm name labels from virt-launcher pod.

Using "hostname" at DHCP options is causing problem with DHCP and the VM end up without IP address

hostname option has to be quoted

# Cannot connect after live-migration

Looks like the endpoint is referencing the old virt-launcher pod wich has the same ip address, removing old virt-launchers fix
the issue
```yaml
apiVersion: v1
kind: Endpoints
metadata:
  annotations:
    endpoints.kubernetes.io/last-change-trigger-time: "2023-02-16T14:24:00Z"
  creationTimestamp: "2023-02-16T14:17:08Z"
  name: ssh-test1-2ghnh
  namespace: clusters-test1
  resourceVersion: "220258"
  uid: c7d4a953-f876-40b9-bfa7-a1ce5d3e43d0
subsets:
- addresses:
  - ip: 10.129.2.25
    nodeName: worker-1
    targetRef:
      kind: Pod
      name: virt-launcher-test1-2ghnh-tkhsb
      namespace: clusters-test1
      uid: d779f8a5-ba16-4146-bb4d-0b489ffee47b
  notReadyAddresses:
  - ip: 10.129.2.25
    nodeName: worker-2
    targetRef:
      kind: Pod
      name: virt-launcher-test1-2ghnh-t2x28
      namespace: clusters-test1
      uid: 650dfac8-00af-4f3f-8c14-c40cc3feb778
  ports:
  - port: 22
    protocol: TCP
```

Sometime the notReadyAddress is not there, looks like ovn-k Recocnile bug that appears when both pods has the same IP

# Nice to have
- [ ] Investigate solutions for VM restart

# Capk demo
https://asciinema.org/a/hEqzoFveibsuHKynQxZ3RxCHZ

**- How to verify it**
This is verify with the umbrella project https://github.com/qinqon/ovn-kubevirt

Docs:
- [ovn + dhcp](https://blog.oddbit.com/post/2019-12-19-ovn-and-dhcp/)
- [Live migration](https://www.youtube.com/watch?v=ijZTMXAg-eI)
- [How to create an Open Virtual Network distributed gateway router](https://developers.redhat.com/blog/2018/11/08/how-to-create-an-open-virtual-network-distributed-gateway-router)
- http://www.openvswitch.org/support/boston2017/1150-ovn-gateways-ipv6.pdf
- [localnet](https://weiti.org/ovn/2018/01/03/ovn-l2-breakout-options.html)
- Load Balancer https://www.cnblogs.com/yaodd/p/7477848.html
- ovn-kuberntes lb:   -  https://github.com/ovn-org/ovn-kubernetes/blob/90d77b70b04530763ed2e8e91bf7a268a25caf42/go-controller/pkg/ovn/loadbalancer/loadbalancer.go#L42
